src/main/java/com/google/idea/perf/Matching.kt

```diff
@@ -192,3 +215,62 @@ private fun fuzzyMatchImpl(source: String, pattern: String): MatchDetails {
 
     return MatchDetails(source, matchedChars, maxScore)
 }
+
+/*
+ * Fuzzy searcher implementation.
+ */
+
+private fun fuzzySearchUncached(
+    sources: Collection<String>,
+    pattern: String,
+    cancellationCheck: () -> Unit
+): List<MatchDetails> {
+    val results = ArrayList<MatchDetails>()
+
+    for ((index, source) in sources.withIndex()) {
+        val match = fuzzyMatchImpl(source, pattern)
+        if (match.matchedChars.size >= pattern.length) {
+            results.add(match)
+        }
+
+        if (index % 256 == 0) {
+            cancellationCheck()
+        }
+    }
+
+    results.sortByDescending { it.score }
+    return results
+}
+
+private typealias ResultCache = LruCache<String, List<String>>
+private typealias PatternCache = LruCache<Collection<String>, ResultCache>
```

Comment:
This cache seems like it could get somewhat large.
IIUC there can be up to 64 `ResultCache` per thread, and up to 1024 `List<String>` per `ResultCache`. And each `List<String>` could have a large size, depending on the query. And each string itself could be long, maybe 50 characters. In the end this could reach GB levels, and that memory is essentially leaked because the `FuzzySearcher` objects survive forever as thread-local variables.

Did you observe that this cache improves performance a lot? Or is it just a hope until we write a benchmark?

Reply:
Yup, the cache size may be too big, I think there is an additional solution to this too. In theory, there are only a few search patterns that yield large `List<String>`s, while the remaining yield small lists. These small lists are vital for the performance of longer search patterns. I think it would be a good idea for me to:

1. Introduce a fallback search method when searching in large lists. This will help flatten out the performance hump of medium-length search patterns for canonical class names.
2. Only add elements in `ResultCache` when the search savings are significant. For example, it makes sense to cache `com.intellij.op`, but not `com.intellij` or `com.`, both of which yield results >50% the size of source list.

I will run this through a memory profiler to verify.